propagation.go

```diff
@@ -41,71 +41,69 @@ var (
 type BuiltinFormat byte
 
 const (
-	// SplitBinary encodes the Span in a SplitBinaryCarrier instance.
+	// Binary encodes the Span in a BinaryCarrier instance.
 	//
-	// The `carrier` for injection and extraction must be a
-	// `*SplitBinaryCarrier` instance.
-	SplitBinary BuiltinFormat = iota
+	// The `carrier` for injection and extraction must be a `BinaryCarrier`
+	// instance (not a pointer to a BinaryCarrier instance).
+	Binary BuiltinFormat = iota
 
-	// SplitText encodes the Span in a SplitTextCarrier instance.
+	// TextMap encodes the Span in a TextMapCarrier instance.
 	//
-	// The `carrier` for injection and extraction must be a `*SplitTextCarrier`
-	// instance.
-	SplitText
+	// The `carrier` for injection and extraction must be a `TextMapCarrier`
+	// instance (not a pointer to a TextMapCarrier instance).
+	TextMap
 
 	// GoHTTPHeader encodes the Span into a Go http.Header instance (both the
 	// tracer state and any baggage).
 	//
 	// The `carrier` for both injection and extraction must be an http.Header
-	// instance.
+	// instance (not a pointer to an http.Header instance).
+	//
+	// If there are entries in the http.Header map prior to a call to Inject(),
+	// they are left alone (i.e., the map is not cleared). Similarly, in calls
+	// to Join() it is fine (and expected in some cases) for the http.Header
+	// map to contain other unrelated data (i.e., non-OpenTracing headers).
 	GoHTTPHeader
+
+	// SplitBinary is DEPRECATED
+	SplitBinary
+	// SplitText is DEPRECATED
+	SplitText
 )
 
-// SplitTextCarrier breaks a propagated Span into two pieces.
+// TextMapCarrier represents a Span for propagation as a key:value map of
+// unicode strings.
 //
-// The Span is separated in this way for a variety of reasons; the most
-// important is to give OpenTracing users a portable way to opt out of
-// Baggage propagation entirely if they deem it a stability risk.
+// If there are entries in the TextMapCarrier prior to a call to Inject(), they
+// are left alone (i.e., the map is not cleared). Similarly, in calls to Join()
+// it is fine (and expected in some cases) for the TextMapCarrier to contain
+// other unrelated data (e.g., arbitrary HTTP header pairs).
+type TextMapCarrier map[string]string
+
+// BinaryCarrier represents a Span for propagation as an opaque byte array.
 //
-// It is legal to provide one or both maps as `nil`; they will be created
-// as needed. If non-nil maps are provided, they will be used without
-// clearing them out on injection.
+// It is fine to pass `nil` to Inject(); in that case, it will allocate a
+// []byte for the caller.
+type BinaryCarrier *[]byte
```

Comment:
I'm fine with this as is. The two main arguments I would have in favor of using `bytes.Buffer` is 1) they're reusable, so user can use a pool of buffers if they want, and 2) `byte.Buffers` are fairly common in Go, so it would remove a couple of steps of turning the `BinaryCarrier` into a buffer post injection in a few cases. Those two arguments are probably premature optimizations though.